lib/core/collections.toit

```diff
@@ -970,24 +970,31 @@ class LargeArray_ extends Array_:
     vector_[1] = Array_ new_size - ARRAYLET_SIZE
     super.from_subclass_
 
-  // An expand that uses the backing arraylets from the old array.  Used by List.
-  resize_for_list_ old_size/int new_size/int:
+  // An expansion or shrinking that uses the backing arraylets from the old
+  // array.  Used by List, Set, and Map.
+  resize_for_list_ copy_size/int new_size/int filler:
     // Rounding up division.
     number_of_arraylets := ((new_size - 1) / ARRAYLET_SIZE) + 1
-    new_vector := Array_ number_of_arraylets
+    if number_of_arraylets == 1:
+      return vector_[0].resize_for_list_ copy_size new_size filler
+    new_vector := Array_ number_of_arraylets  // new_vector may be a LargeArray_!
     left := new_size
+    pos := 0
     number_of_arraylets.repeat:
-      if (it + 1) * ARRAYLET_SIZE <= old_size:
-        new_vector[it] = vector_[it]
-        left -= ARRAYLET_SIZE
+      arraylet := ?
+      limit := min left ARRAYLET_SIZE
+      if pos + limit <= size:
+        arraylet = vector_[it]
+        if pos + limit > copy_size:
```

Comment:
Is this check necessary and worth it?

Reply:
It's hard to measure a difference.  I took it out.